### PR TITLE
Remote Schema Permissions (Role based only)

### PR DESCRIPTION
**Pull request by codingkarthik**
PR for Role based remote schema permissions, [RFC](https://github.com/tirumaraiselvan/graphql-engine/blob/rfc-remote-schema-perms/rfcs/remote-schema-permissions.md)

Blocked on: 

https://github.com/hasura/graphql-parser-hs/pull/36

TODO:

- [x] Feature flag for enabling remote schema permissions
- [x] Tests
- [x] docs
- [] Changelog (added a temporary one for now)

### Changelog

- [ ] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Server
- [ ] Console
- [ ] CLI
- [x] Docs
- [ ] Community Content
- [ ] Build System
- [x] Tests
- [ ] Other (list it)

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [ ] No
- [x] Yes
  - [ ] Updated docs with SQL for downgrading the catalog 

#### Metadata

Does this PR add a new Metadata feature?
- [ ] No
- [x] Yes
  - Does `run_sql` auto manages the new metadata through schema diffing?
    - [ ] Yes
    - [ ] Not required
  - Does `run_sql` auto manages the definitions of metadata on renaming?
    - [ ] Yes
    - [ ] Not required
  - Does `export_metadata`/`replace_metadata` supports the new metadata added?
    - [x] Yes
    - [ ] Not required

#### GraphQL
- [ ] No new GraphQL schema is generated
- [ ] New GraphQL schema is being generated:
   - [ ] New types and typenames are correlated
   
   

#### Breaking changes

- [ ] No Breaking changes
- [ ] There are breaking changes:

  1. Metadata API

     Existing `query` types:
     - [ ] Modify `args` payload which is not backward compatible
     - [ ] Behavioural change of the API
     - [ ] Change in response `JSON` schema
     - [ ] Change in error code
     

  2. GraphQL API

     Schema Generation:
     
     
     - [ ] Change in any `NamedType`
     - [ ] Change in table field names
     

     Schema Resolve:-
     
     - [ ] Change in treatment of `null` value for any input fields 
     

  3. Logging

     - [ ] Log `JSON` schema has changed
     - [ ] Log `type` names have changed

### CODE REVIEW

File: docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+   :backlinks: none
+   :depth: 1
+   :local:
+
+Introduction
+------------
+
+The permission layer is designed to restrict the operations that can be
+performed by various users. Permissions can be defined to only expose a certain
+part of the upstram remote schema for each role. By default, the ``admin`` role
```
Comment (soorajshankar) [inline docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst:20]:
```suggestion
part of the upstream remote schema for each role. By default, the ``admin`` role
```

File: docs/graphql/core/api-reference/schema-metadata-api/index.rst [hunk 2]
```diff
@@ -458,6 +469,7 @@ See :ref:`server_flag_reference` for info on setting the above flag/env var.
   Custom Functions <custom-functions>
   Relationships <relationship>
   Permissions <permission>
+  Remote Schema Permission <remote-schema-permissions>
```
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/index.rst:472]:
```suggestion
  Remote Schema Permissions <remote-schema-permissions>
```

File: docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+has unrestricted access to the upstream remote schema.
+
+.. _add_remote_schema_permissions:
+
+add_remote_schema_permissions
+-----------------------------
+
+A remote schema permission is used to enforce constraints on the query that's
```
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst:28]:
```suggestion
A remote schema permission is used to enforce constraints on queries
```

File: docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+being queried to the remote schema. This is done by specifying the schema document
```
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst:29]:
```suggestion
to the remote schema. This is done by specifying the schema document
```

File: docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+  schema {
+    query: Query
+    mutation: mutation_root
+  }
+
+Let's say we want to impose some restrictions on the ``user`` role:
+
+1. Expose only the ``user_id``, ``name`` and the ``user_messages`` field in the ``User`` object.
+2. Allow filtering of the messages only by ``name`` in the ``whered`` argument
```
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst:95]:
```suggestion
2. Allow filtering of the messages only by ``name`` in the ``where`` argument
```

File: docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+   of the ``user_messages`` field.
+3. Allow only ``hello``, ``messages`` and the ``user`` top level node in the ``Query`` object.
```
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst:97]:
```suggestion
3. Allow only the fields ``hello``, ``messages`` and the ``user`` top level node in the ``Query`` object.
```

File: docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+     - :ref:`RemoteSchemaName`
+     - Name of the remote schema
+   * - role
+     - true
+     - :ref:`RoleName`
+     - Role
+   * - definition
+     - true
+     - RemoteSchemaPermission_
+     - The permission definition
```
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst:195]:
```suggestion
     - The remote schema permission definition
```

File: docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+     - Description
+   * - schema
+     - true
+     - GraphQL SDL
+     - GraphQL SDL defining the role based schema
+
+.. note::
+   ``add_remote_schema_permissions`` will only work when the graphql-engine has enabled remote
+   schema permissions. Remote schema permissions can be enabled by running the graphql-engine
+   with ``--enable-remote-schema-permissions`` server flag or by setting the   ``HASURA_GRAPHQL_ENABLE_REMOTE_SCHEMA_PERMISSIONS`` environment variable.
```
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst:221]:
```suggestion
   with the ``--enable-remote-schema-permissions`` server flag or by setting the   ``HASURA_GRAPHQL_ENABLE_REMOTE_SCHEMA_PERMISSIONS`` environment variable.
```

**Review by marionschleifer — CHANGES_REQUESTED** [review]

**Comment by codingkarthik** [comment]
Closing this PR, the features of this PR are included in https://github.com/hasura/graphql-engine/pull/6079

**Pull request closed without merging**